src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Configuration.Settings;
+using Stratis.Bitcoin.P2P.Peer;
+using Stratis.Bitcoin.Utilities;
+using Stratis.Bitcoin.Utilities.Extensions;
+
+namespace Stratis.Bitcoin.Consensus
+{
+    public interface IChainedHeaderValidator
+    {
+        void Validate(ChainedHeader chainedHeader);
+    }
+
+    /// <summary>
+    /// A tree structure of <see cref="ChainedHeader "/> elements.
+    /// </summary>
+    /// <remarks>
+    /// A blockchain can have multiple versions of the chain, with only one being the longest chain (or the chain with most proof).
+    /// While this is not ideal it can happen naturally or maliciously, to be able to find the best chain we have to keep track of all chains we discover.
```

Comment:
All chains that don't violate the max reorg rule (when it's enabled) and might potentially become better than our CT.